File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.cpp
class Solution {
public:
    int count = 0;

    void backtrack(vector<int> &freq) {
        for (int i = 0; i < 26; i++) {
            if (freq[i] > 0) {
                count++;   // Every valid choice is a new sequence
                
                freq[i]--; // Use this tile
                backtrack(freq);
                freq[i]++; // Restore for other choices
            }
        }
    }

    int numTilePossibilities(string tiles) {
        vector<int> freq(26, 0);
        
        for (char ch : tiles)
            freq[ch - 'A']++;  // Count tile frequencies
        
        backtrack(freq);
        return count;
    }
};

File: 1160-letter-tile-possibilities/1160-letter-tile-possibilities.java
import java.util.*;

class Solution {
    int count = 0;

    public int numTilePossibilities(String tiles) {
        Map<Character, Integer> map = new HashMap<>();

        // Count the frequency of each character
        for (char c : tiles.toCharArray()) {
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        // Start the backtracking process
        solve(map);
        return count;
    }

    public void solve(Map<Character, Integer> map) {
        // For each character, if its count is > 0, it can be used to form a sequence
        for (Map.Entry<Character, Integer> i : map.entrySet()) {
            // If the character count is positive
            if (i.getValue() > 0)
             {
                // Form a valid sequence by using this character
                count++; // Add this sequence

                // Reduce the frequency of this character
                map.put(i.getKey(), i.getValue() - 1);

                // Continue to form further sequences
                solve(map);

                // Backtrack, restore the frequency of the character
                map.put(i.getKey(), i.getValue() + 1);
            }
        }
    }
}
